### PR TITLE
Added Configuration information to Blockchain

### PR DESCRIPTION
**Pull request by fishseabowl**
1. Add configuration information to blockchain
2. define genesis block struct
3. Add unit test

### CODE REVIEW

File: pyrsia_node/src/main.rs [hunk 1]
```diff
@@ -196,7 +198,7 @@ async fn main() {
     tokio::spawn(server);
     let tx4 = tx.clone();
 
-    let raw_chain = block_chain::Blockchain::new();
+    let raw_chain = block_chain::Blockchain::new(&local_key);
```
Comment (mgrand) [inline pyrsia_node/src/main.rs:201]: You should not have functions over 100 lines long. Breaking this into sub-functions will make this more readable, easier for you to find bugs and if this is ever maintained by someone else, easier for them to understand. Especially if the sub-functions have meaningful names.

Most IDEs have a way to automate the refactoring of a chunk of code into a separate function.

File: src/block_chain/block.rs [hunk 2]
```diff
@@ -38,7 +38,7 @@ impl Signature {
 }
 
 type TransactionSignature = Signature;
-type BlockSignature = Signature;
+pub type BlockSignature = Signature;
 
 pub fn sign(msg: &[u8], keypair: &identity::ed25519::Keypair) -> Vec<u8> {
```
Comment (mgrand) [inline src/block_chain/block.rs:43]: What is being signed here. It is not obvious. If it is the block, you will need to add a way to securely manage keys. If it is the transaction, that is a problem. Transactions should be signed before they get to the blockchain.

File: src/block_chain/block_chain.rs [hunk 3]
```diff
@@ -13,91 +13,157 @@
...
 use libp2p::identity;
 use rand::Rng;
 use serde::{Deserialize, Serialize};
 use std::fmt::{Display, Formatter};
 
+/// BlockchainId identifies the current chain
 #[derive(Serialize, Deserialize, Debug, Clone)]
-pub struct Blockchain {
-    pub blocks: Vec<Block>,
+pub enum BlockchainId {
```
Comment (mgrand) [inline src/block_chain/block_chain.rs:26]: Unclear what this is for.

File: src/block_chain/block_chain.rs [hunk 4]
```diff
@@ -13,91 +13,157 @@
...
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct Blockchain {
+    pub genesis_block: GenesisBlock,
+    pub blocks: Vec<Block>,
+}
+
+impl Blockchain {
+    pub fn new(keypair: &identity::Keypair) -> Self {
+        let ed25519_keypair = match keypair {
```
Comment (mgrand) [inline src/block_chain/block_chain.rs:106]: A better way of structuring this is to have Keypair be an enum and put the logic for operations specified to the type of keypair in the impl for the keypair. If you want to see an example, look at what I do with Hash in artifact manager.

**Review by mgrand — COMMENTED** [review]
It is difficult to give any high-level comments, because I do not know what you are trying to accomplish. It would be very helpful to have in the PR a link to a document that explains what you are trying to accomplish overall. If you have not written such a document (it doesn't need to be long), you may find the writing it will help clarify your own thoughts.

File: pyrsia_node/src/main.rs (hunk 1 above)
Comment (prince-chrismc) [reply pyrsia_node/src/main.rs:201]: There's only three lines of diff here, feel free to open a separate issue 👍

File: src/block_chain/block_chain.rs [hunk 5]
```diff
@@ -13,91 +13,157 @@
...
         keypair: &identity::ed25519::Keypair,
         transactions: &[Transaction],
     ) {
-        let last_block = match self.blocks.last() {
-            Some(block) => block,
-            None => {
-                Blockchain::genesis(self, keypair);
-                return;
-            }
+        let (parent_hash, previous_number) = match self.blocks.last() {
```
Comment (prince-chrismc) [inline src/block_chain/block_chain.rs:122]: You can just add the genesis block when you initialize the vector [here](https://github.com/pyrsia/pyrsia/pull/295/files#diff-ad2b62fe242505ea5bceca51cd0d423999a3aa8de6515c426aa166aff4d9a505R112-R113) and just assert the size is >= 1

File: src/block_chain/block_chain.rs (hunk 3 above)
Comment (prince-chrismc) [reply src/block_chain/block_chain.rs:26]: It's hard to image Pyrsia having more then one blockchain at this time 🤔

File: Cargo.lock [hunk 6]
```diff
@@ -3642,6 +3642,7 @@ dependencies = [
 [[package]]
 name = "sha1"
 version = "0.6.1"
+
```
Comment (prince-chrismc) [inline Cargo.lock:3645]: Seem out of place 🤔

File: src/block_chain/block.rs [hunk 7]
```diff
@@ -151,9 +151,49 @@ pub fn verify(pubkey: &identity::ed25519::PublicKey, msg: &[u8], sig: &[u8]) ->
...
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use rand::Rng;
+
+    #[test]
+    fn test_build_block() -> Result<(), String> {
+        let keypair = identity::ed25519::Keypair::generate();
```
Comment (prince-chrismc) [inline src/block_chain/block.rs:173]: You expose a helper [here](https://github.com/pyrsia/pyrsia/pull/295/files#diff-b34f72ca0859bced09bf9b55c159ec83f584736180ab108626113dbc22072af5R154) but don't use it... was that intentional?

**Review by prince-chrismc — DISMISSED** [review]
I like the idea of the config being saved in the genesis block.

Bunch of editorial comments from me but this is enough of an improvement we can save refactoring for the next PR.

Lots of good comments and questions from the team so try answering a few 🙏

File: pyrsia_node/src/main.rs (hunk 1 above)
Comment (fishseabowl) [reply pyrsia_node/src/main.rs:201]: Sure. I would refactor communication methods in the next PR.

File: src/block_chain/block.rs (hunk 2 above)
Comment (fishseabowl) [reply src/block_chain/block.rs:43]: Thanks for your comments. This is the temporary solution. I would bind these methods with related struct.

File: src/block_chain/block.rs (hunk 7 above)
Comment (fishseabowl) [reply src/block_chain/block.rs:173]: Sorry. What are your concerns? Display trait or unit test?

File: src/block_chain/block_chain.rs (hunk 3 above)
Comment (fishseabowl) [reply src/block_chain/block_chain.rs:26]: Test networks, different consensus algorithms, or different contents in the ledger.

File: src/block_chain/block_chain.rs (hunk 4 above)
Comment (fishseabowl) [reply src/block_chain/block_chain.rs:106]: It is currently used for easy serialization and deserialization. I would continue to refactor these codes.

File: src/block_chain/block_chain.rs (hunk 5 above)
Comment (fishseabowl) [reply src/block_chain/block_chain.rs:122]: In my definition, a genesis block is not a block. There are different data structs.

File: Cargo.lock (hunk 6 above)
Comment (fishseabowl) [reply Cargo.lock:3645]: Let me check. I don't use sha1 algorithm.

Comment (mgrand) [reply Cargo.lock:3645]: Sha1 is used by Synapse for Bittorrent. V1 torrents use sha1. V2 supports SHA256. Neither Synapse nor any of the Rust based bittorrent implementations that I have found support v2. That is something to be addressed in the future. We will be compensating by having the artifact manager rehash the files using a stronger hash before allowing them to be pulled.

**Review by prince-chrismc — APPROVED** [review]

File: src/block_chain/block.rs (hunk 7 above)
Comment (prince-chrismc) [reply src/block_chain/block.rs:173]: Just some odd code duplication/unused, nothing critical

File: src/block_chain/block_chain.rs (hunk 5 above)
Comment (prince-chrismc) [reply src/block_chain/block_chain.rs:122]: Thats very interesting. 🤔

**Review by chb0github — APPROVED** [review]

**Pull request merged**